core/include/bitarithm.h

```diff
@@ -164,6 +164,37 @@ static inline unsigned bitarithm_lsb(unsigned v)
 }
 #endif
 
+/**
+ * @brief   Returns the index of the first set bit in @p state, returns @p state with
+ *          that bit cleared.
+ *
+ * @warning This is a low-level helper function, arguments are not checked.
+ *          It is intended to iterate over a bit map until all bits are cleared.
+ *          Whether it starts with the highest or lowest bit will depend on what is fastest
+ *          on the given hardware.
+ *
+ * @param[in]   state   Bit Map with at least one bit set
+ * @param[out]  index   Index of the first set bit. Must be initialized with 0 before the
+ *                      first call to this function, must not be modified between subsequent
+ *                      calls.
+ *
+ * @return      new state value - must be treated as opaque value
+ *
+ */
+static inline unsigned bitarithm_test_and_clear(unsigned state, uint8_t *index)
+{
+#if defined(BITARITHM_HAS_CLZ)
+    *index = 8 * sizeof(state) - __builtin_clz(state) - 1;
+    return state & ~(1 << *index);
+#else
+    while ((state & 1) == 0) {
+        *index += 1;
+        state >>= 1;
+    }
+    return state & ~1;
+#endif
+}
```

Comment:
I'm wondering if it makes sense to reuse `bitharithm_lsb here. That one also has a lookup-based implementation in case the builtin is not available.

Reply:
I wouldn't use `bitarithm_lsb` directly as it falls back to shifts too if `BITARITHM_LSB_LOOKUP` is not set, but we can re-use the algorithm here and re-use the intermediate value too.

Gives a nice speedup

```
+ bitarithm_test_and_clear: 323232 iterations per second
```

Reply:
> `+ bitarithm_test_and_clear: 323232 iterations per second`

It's faster than the built-in function? :thinking:

Reply:
On `stm32f429i-disco` which does have `clz`, I get

#### __builtin_clz

```
+ bitarithm_test_and_clear: 2382974 iterations per second
```

#### De Bruijn
```
+ bitarithm_test_and_clear: 1592414 iterations per second
```

#### Shifts
```
+ bitarithm_test_and_clear: 539325 iterations per second
```

Reply:
Ah, that definitely makes more sense, thanks :)